\documentclass{article}\usepackage[]{graphicx}\usepackage[]{color}
%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\definecolor{fgcolor}{rgb}{0.345, 0.345, 0.345}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.686,0.059,0.569}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.192,0.494,0.8}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.678,0.584,0.686}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.345,0.345,0.345}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.161,0.373,0.58}{\textbf{#1}}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0.69,0.353,0.396}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.333,0.667,0.333}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.737,0.353,0.396}{\textbf{#1}}}%
\let\hlipl\hlkwb

\usepackage{framed}
\makeatletter
\newenvironment{kframe}{%
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\definecolor{shadecolor}{rgb}{.97, .97, .97}
\definecolor{messagecolor}{rgb}{0, 0, 0}
\definecolor{warningcolor}{rgb}{1, 0, 1}
\definecolor{errorcolor}{rgb}{1, 0, 0}
\newenvironment{knitrout}{}{} % an empty environment to be redefined in TeX

\usepackage{alltt}
\usepackage{underscore}
\usepackage[utf8]{inputenc}
\usepackage{longtable}
\usepackage[margin=1in]{geometry}
%\usepackage[spanish]{babel}
\usepackage{hyperref}
\usepackage{graphicx}
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\begin{document}

\title{A quick gene selection, annotation and GO analysis}
\author{Alex S\'anchez$\sp {1,2}$}
\date{$\sp 1$Statistics and Bioinformatics Unit (VHIR)\\
 $\sp 2$ Statistics Department. University of Barcelona}

\maketitle

\tableofcontents






\section{Introduction}

Most gene expression studies undergo one phase where, after gene
selection has been performed, one wishes to:
\begin{enumerate}
\item Annotate the genes or transcripts, that is associate, to each
  probeset or transcript, some identifiers in the appropriate
  databases that can be used to understand better the results or that
  are needed to proceed with further analyses (for instance GO
  Analysis needs ``Entrez'' identifiers).
\item Do some type of Gene Set Enrichment Analyses such as
  Overrepresentation Analysis (ORA) or classical Gene Set Enrichment
  Analysis (GSEA).
\end{enumerate}

This document is an illustration which does not intend to be exhaustive, on how to do this with some of these packages.

\subsection{Obtaining gene lists}

The first step in annotation analysis is to obtain the gene lists,
usually as the output of some differential expression analysis.

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{topTab} \hlkwb{<-} \hlkwd{read.table}\hlstd{(}\hlstr{"https://raw.githubusercontent.com/alexsanchezpla/Ejemplo_de_MDA_con_Bioconductor/master/results/ExpressAndTop_AvsB.csv2"}\hlstd{,} \hlkwc{head}\hlstd{=}\hlnum{TRUE}\hlstd{,} \hlkwc{sep}\hlstd{=}\hlstr{";"}\hlstd{,} \hlkwc{dec}\hlstd{=}\hlstr{","}\hlstd{,} \hlkwc{row.names} \hlstd{=} \hlnum{1}\hlstd{)}
\hlkwd{colnames}\hlstd{(topTab)}
\end{alltt}
\begin{verbatim}
##  [1] "SymbolsA"     "EntrezsA"     "logFC"        "AveExpr"      "t"           
##  [6] "P.Value"      "adj.P.Val"    "B"            "GSM26878.CEL" "GSM26883.CEL"
## [11] "GSM26887.CEL" "GSM26903.CEL" "GSM26910.CEL" "GSM26888.CEL" "GSM26889.CEL"
## [16] "GSM26892.CEL" "GSM26898.CEL" "GSM26906.CEL"
\end{verbatim}
\begin{alltt}
\hlkwd{head}\hlstd{(topTab)}
\end{alltt}
\begin{verbatim}
##             SymbolsA EntrezsA  logFC AveExpr       t          P.Value
## 204667_at      FOXA1     3169 -3.038   8.651 -14.362 0.00000000005742
## 215729_s_at    VGLL1    51442  3.452   6.138  12.815 0.00000000034385
## 220192_x_at    SPDEF    25803 -3.016   9.522 -10.859 0.00000000433617
## 214451_at     TFAP2B     7021 -5.665   7.433 -10.830 0.00000000451614
## 217528_at      CLCA2     9635 -5.622   6.763  -9.666 0.00000002429965
## 217284_x_at   SERHL2   253190 -4.313   9.133  -9.528 0.00000002994504
##               adj.P.Val      B GSM26878.CEL GSM26883.CEL GSM26887.CEL
## 204667_at   0.000000357 14.650        9.822        9.514        9.919
## 215729_s_at 0.000001069 13.150        4.737        4.761        6.255
## 220192_x_at 0.000007020 10.929       10.484       10.915       10.511
## 214451_at   0.000007020 10.893       10.177       10.060       11.201
## 217528_at   0.000030219  9.364       10.534       10.036       11.326
## 217284_x_at 0.000031033  9.172       11.727        9.741       11.436
##             GSM26903.CEL GSM26910.CEL GSM26888.CEL GSM26889.CEL GSM26892.CEL
## 204667_at          9.601        9.592        6.484        6.551        7.001
## 215729_s_at        4.820        4.848        8.266        8.963        8.304
## 220192_x_at       11.510       10.265        7.824        7.810        7.522
## 214451_at         10.889       10.404        4.818        4.784        4.976
## 217528_at          8.053       10.619        4.581        4.538        4.519
## 217284_x_at       12.819       12.687        7.274        7.298        7.491
##             GSM26898.CEL GSM26906.CEL
## 204667_at          6.685        6.535
## 215729_s_at        8.769        8.381
## 220192_x_at        8.427        7.020
## 214451_at          4.912        4.916
## 217528_at          4.357        4.463
## 217284_x_at        7.562        7.217
\end{verbatim}
\end{kframe}
\end{knitrout}

\section{Annotating the genes}

This table has already been ``annotated'' in the script that has
performed the original analysis, but, \emph{what would we have had to do if
it hadn't been}?

We might have used either a specific annotation package for the array
or the BioMaRt package.

\subsection{Using a microarray annotation package}

If we hadn't had 'Entrez' Identifiers, but only the probeset identifiers which depend on the array type we might have done as follows:

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{probeIDsAll} \hlkwb{<-} \hlkwd{rownames}\hlstd{(topTab)}
\hlstd{probeIDsUp} \hlkwb{<-} \hlstd{probeIDsAll [topTab}\hlopt{$}\hlstd{adj.P.Val}\hlopt{<}\hlnum{0.05} \hlopt{&} \hlstd{topTab}\hlopt{$}\hlstd{logFC} \hlopt{>} \hlnum{0}\hlstd{]}
\hlstd{probeIDsDown} \hlkwb{<-} \hlstd{probeIDsAll [topTab}\hlopt{$}\hlstd{adj.P.Val}\hlopt{<}\hlnum{0.05} \hlopt{&} \hlstd{topTab}\hlopt{$}\hlstd{logFC} \hlopt{<} \hlnum{0}\hlstd{]}

\hlkwd{require}\hlstd{(hgu133a.db)}
\hlkwd{keytypes}\hlstd{(hgu133a.db)}
\end{alltt}
\begin{verbatim}
##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GO"           "GOALL"        "IPI"          "MAP"          "OMIM"        
## [16] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROBEID"      "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
## [26] "UNIGENE"      "UNIPROT"
\end{verbatim}
\begin{alltt}
\hlstd{geneEntrezsUp} \hlkwb{<-} \hlkwd{select}\hlstd{(hgu133a.db,} \hlkwc{keys} \hlstd{= probeIDsUp,} \hlkwc{columns}\hlstd{=}\hlkwd{c}\hlstd{(}\hlstr{"ENTREZID"}\hlstd{,} \hlstr{"SYMBOL"}\hlstd{))}
\end{alltt}


{\ttfamily\noindent\itshape\color{messagecolor}{\#\# 'select()' returned 1:1 mapping between keys and columns}}\begin{alltt}
\hlstd{geneEntrezsDown} \hlkwb{<-} \hlkwd{select}\hlstd{(hgu133a.db,} \hlkwc{keys} \hlstd{= probeIDsUp,} \hlkwc{columns}\hlstd{=}\hlkwd{c}\hlstd{(}\hlstr{"ENTREZID"}\hlstd{,} \hlstr{"SYMBOL"}\hlstd{))}
\end{alltt}


{\ttfamily\noindent\itshape\color{messagecolor}{\#\# 'select()' returned 1:1 mapping between keys and columns}}\begin{alltt}
\hlstd{geneEntrezsUniverse} \hlkwb{<-} \hlkwd{select}\hlstd{(hgu133a.db,} \hlkwc{keys} \hlstd{= probeIDsAll,} \hlkwc{columns}\hlstd{=}\hlkwd{c}\hlstd{(}\hlstr{"ENTREZID"}\hlstd{,} \hlstr{"SYMBOL"}\hlstd{))}
\end{alltt}


{\ttfamily\noindent\itshape\color{messagecolor}{\#\# 'select()' returned 1:1 mapping between keys and columns}}\begin{alltt}
\hlkwd{head}\hlstd{(geneEntrezsUp)}
\end{alltt}
\begin{verbatim}
##       PROBEID ENTREZID SYMBOL
## 1 215729_s_at    51442  VGLL1
## 2   205044_at     2568  GABRP
## 3   209337_at    11168  PSIP1
## 4   209786_at    10473  HMGN4
## 5   204061_at     5613   PRKX
## 6   207039_at     1029 CDKN2A
\end{verbatim}
\end{kframe}
\end{knitrout}

\subsection{Using BiomaRt}

Biomart is a powerful engine for linking identifiers.
It is a bit cryptic at the first approach because in order to use it we must define \emph{filters} (what we input for searching), \emph{attributes} (what we output) and \emph{values} (which values we input).

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{biodataset} \hlkwb{<-} \hlkwd{useDataset}\hlstd{(}\hlstr{"hsapiens_gene_ensembl"}\hlstd{,} \hlkwd{useMart}\hlstd{(}\hlstr{"ensembl"}\hlstd{))}
\hlkwd{listDatasets}\hlstd{(biodataset)}\hlopt{$}\hlstd{dataset}
\end{alltt}
\begin{verbatim}
##   [1] "acalliptera_gene_ensembl"       "acarolinensis_gene_ensembl"    
##   [3] "acitrinellus_gene_ensembl"      "amelanoleuca_gene_ensembl"     
##   [5] "amexicanus_gene_ensembl"        "anancymaae_gene_ensembl"       
##   [7] "aocellaris_gene_ensembl"        "apercula_gene_ensembl"         
##   [9] "aplatyrhynchos_gene_ensembl"    "apolyacanthus_gene_ensembl"    
##  [11] "atestudineus_gene_ensembl"      "btaurus_gene_ensembl"          
##  [13] "caperea_gene_ensembl"           "catys_gene_ensembl"            
##  [15] "ccapucinus_gene_ensembl"        "cchok1gshd_gene_ensembl"       
##  [17] "ccrigri_gene_ensembl"           "celegans_gene_ensembl"         
##  [19] "cfamiliaris_gene_ensembl"       "chircus_gene_ensembl"          
##  [21] "choffmanni_gene_ensembl"        "cintestinalis_gene_ensembl"    
##  [23] "cjacchus_gene_ensembl"          "clanigera_gene_ensembl"        
##  [25] "cpalliatus_gene_ensembl"        "cporcellus_gene_ensembl"       
##  [27] "csabaeus_gene_ensembl"          "csavignyi_gene_ensembl"        
##  [29] "csemilaevis_gene_ensembl"       "csyrichta_gene_ensembl"        
##  [31] "cvariegatus_gene_ensembl"       "dmelanogaster_gene_ensembl"    
##  [33] "dnovemcinctus_gene_ensembl"     "dordii_gene_ensembl"           
##  [35] "drerio_gene_ensembl"            "eburgeri_gene_ensembl"         
##  [37] "ecaballus_gene_ensembl"         "eeuropaeus_gene_ensembl"       
##  [39] "elucius_gene_ensembl"           "etelfairi_gene_ensembl"        
##  [41] "falbicollis_gene_ensembl"       "fcatus_gene_ensembl"           
##  [43] "fdamarensis_gene_ensembl"       "fheteroclitus_gene_ensembl"    
##  [45] "gaculeatus_gene_ensembl"        "gaffinis_gene_ensembl"         
##  [47] "ggallus_gene_ensembl"           "ggorilla_gene_ensembl"         
##  [49] "gmorhua_gene_ensembl"           "hburtoni_gene_ensembl"         
##  [51] "hcomes_gene_ensembl"            "hfemale_gene_ensembl"          
##  [53] "hmale_gene_ensembl"             "hsapiens_gene_ensembl"         
##  [55] "ipunctatus_gene_ensembl"        "itridecemlineatus_gene_ensembl"
##  [57] "jjaculus_gene_ensembl"          "kmarmoratus_gene_ensembl"      
##  [59] "lafricana_gene_ensembl"         "lbergylta_gene_ensembl"        
##  [61] "lchalumnae_gene_ensembl"        "loculatus_gene_ensembl"        
##  [63] "malbus_gene_ensembl"            "marmatus_gene_ensembl"         
##  [65] "mauratus_gene_ensembl"          "mcaroli_gene_ensembl"          
##  [67] "mdomestica_gene_ensembl"        "mfascicularis_gene_ensembl"    
##  [69] "mfuro_gene_ensembl"             "mgallopavo_gene_ensembl"       
##  [71] "mleucophaeus_gene_ensembl"      "mlucifugus_gene_ensembl"       
##  [73] "mmola_gene_ensembl"             "mmulatta_gene_ensembl"         
##  [75] "mmurinus_gene_ensembl"          "mmusculus_gene_ensembl"        
##  [77] "mnemestrina_gene_ensembl"       "mochrogaster_gene_ensembl"     
##  [79] "mpahari_gene_ensembl"           "mspretus_gene_ensembl"         
##  [81] "mzebra_gene_ensembl"            "nbrichardi_gene_ensembl"       
##  [83] "neugenii_gene_ensembl"          "ngalili_gene_ensembl"          
##  [85] "nleucogenys_gene_ensembl"       "oanatinus_gene_ensembl"        
##  [87] "oaries_gene_ensembl"            "ocuniculus_gene_ensembl"       
##  [89] "odegus_gene_ensembl"            "ogarnettii_gene_ensembl"       
##  [91] "ohni_gene_ensembl"              "ohsok_gene_ensembl"            
##  [93] "olatipes_gene_ensembl"          "omelastigma_gene_ensembl"      
##  [95] "oniloticus_gene_ensembl"        "oprinceps_gene_ensembl"        
##  [97] "pabelii_gene_ensembl"           "paltaica_gene_ensembl"         
##  [99] "panubis_gene_ensembl"           "pbairdii_gene_ensembl"         
## [101] "pcapensis_gene_ensembl"         "pcoquereli_gene_ensembl"       
## [103] "pformosa_gene_ensembl"          "pkingsleyae_gene_ensembl"      
## [105] "platipinna_gene_ensembl"        "pmagnuspinnatus_gene_ensembl"  
## [107] "pmarinus_gene_ensembl"          "pmexicana_gene_ensembl"        
## [109] "pnattereri_gene_ensembl"        "pnyererei_gene_ensembl"        
## [111] "ppaniscus_gene_ensembl"         "ppardus_gene_ensembl"          
## [113] "preticulata_gene_ensembl"       "psinensis_gene_ensembl"        
## [115] "ptroglodytes_gene_ensembl"      "pvampyrus_gene_ensembl"        
## [117] "rbieti_gene_ensembl"            "rnorvegicus_gene_ensembl"      
## [119] "rroxellana_gene_ensembl"        "saraneus_gene_ensembl"         
## [121] "sboliviensis_gene_ensembl"      "scerevisiae_gene_ensembl"      
## [123] "sdorsalis_gene_ensembl"         "sdumerili_gene_ensembl"        
## [125] "sformosus_gene_ensembl"         "sharrisii_gene_ensembl"        
## [127] "smaximus_gene_ensembl"          "spartitus_gene_ensembl"        
## [129] "sscrofa_gene_ensembl"           "tbelangeri_gene_ensembl"       
## [131] "tguttata_gene_ensembl"          "tnigroviridis_gene_ensembl"    
## [133] "trubripes_gene_ensembl"         "ttruncatus_gene_ensembl"       
## [135] "vpacos_gene_ensembl"            "xcouchianus_gene_ensembl"      
## [137] "xmaculatus_gene_ensembl"        "xtropicalis_gene_ensembl"
\end{verbatim}
\begin{alltt}
\hlstd{filters}\hlkwb{<-}\hlkwd{listFilters}\hlstd{(biodataset)}
\hlcom{# We need to find the filter to link with Affymetrx arrays hgu133a}
\hlstd{u133aFilters}\hlkwb{<-} \hlkwd{grep}\hlstd{(}\hlstr{"u133a"}\hlstd{, filters[,}\hlnum{1}\hlstd{] )}
\hlstd{u133aFilters} \hlkwb{<-} \hlstd{filters[u133aFilters,]}
\hlstd{myu133aFilter} \hlkwb{<-} \hlstd{u133aFilters[}\hlnum{3}\hlstd{,}\hlnum{1}\hlstd{]}
\hlstd{myu133aFilter}
\end{alltt}
\begin{verbatim}
## [1] "affy_hg_u133a"
\end{verbatim}
\begin{alltt}
\hlstd{atributs}\hlkwb{<-} \hlkwd{listAttributes}\hlstd{(biodataset)}
\hlstd{entrezAtributs}\hlkwb{<-} \hlkwd{grep}\hlstd{(}\hlstr{"entrez"}\hlstd{, atributs[,}\hlnum{1}\hlstd{])}
\hlstd{entrezAtribut} \hlkwb{<-} \hlstd{atributs[entrezAtributs,]}
\hlstd{myentrezAtribut} \hlkwb{<-} \hlstd{entrezAtribut[}\hlnum{2}\hlstd{,}\hlnum{1}\hlstd{]}
\hlstd{myentrezAtribut}
\end{alltt}
\begin{verbatim}
## [1] "entrezgene"
\end{verbatim}
\begin{alltt}
\hlcom{# Now we can do the search}
\hlstd{entrezfromProbesUp} \hlkwb{<-} \hlkwd{getBM}\hlstd{(}\hlkwc{filters}\hlstd{= myu133aFilter,}
                          \hlkwc{attributes}\hlstd{=} \hlkwd{c}\hlstd{(myentrezAtribut, myu133aFilter),}
                          \hlkwc{values}\hlstd{= probeIDsUp,}
                          \hlkwc{mart}\hlstd{= biodataset,}\hlkwc{uniqueRows}\hlstd{=}\hlnum{TRUE}\hlstd{)}
\hlkwd{head}\hlstd{(entrezfromProbesUp)}
\end{alltt}
\begin{verbatim}
##   entrezgene affy_hg_u133a
## 1         NA     202396_at
## 2      29078     219006_at
## 3       6515   202499_s_at
## 4  100529262   206560_s_at
## 5      51389   219598_s_at
## 6       6877     210053_at
\end{verbatim}
\end{kframe}
\end{knitrout}

\subsection{The gene list for pathway Analysis}

In this example we had already had the Entrez and Symbol identifiers so we can extract these directly from the topTable.

Although we skip it here it may be interesting to compare the entrez identifiers obtained from the three distinct approaches. They should be identical, but there may be small discrepancies...

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{geneListUp} \hlkwb{<-} \hlstd{topTab}\hlopt{$}\hlstd{EntrezsA [topTab}\hlopt{$}\hlstd{adj.P.Val}\hlopt{<}\hlnum{0.05} \hlopt{&} \hlstd{topTab}\hlopt{$}\hlstd{logFC} \hlopt{>} \hlnum{0}\hlstd{]}
\hlkwd{head}\hlstd{(geneListUp)}
\end{alltt}
\begin{verbatim}
## [1] 51442  2568 11168 10473  5613  1029
\end{verbatim}
\begin{alltt}
\hlstd{geneListDown} \hlkwb{<-} \hlstd{topTab}\hlopt{$}\hlstd{EntrezsA [topTab}\hlopt{$}\hlstd{adj.P.Val}\hlopt{<}\hlnum{0.05} \hlopt{&} \hlstd{topTab}\hlopt{$}\hlstd{logFC} \hlopt{<} \hlnum{0}\hlstd{]}
\hlkwd{length}\hlstd{(geneListDown)}
\end{alltt}
\begin{verbatim}
## [1] 268
\end{verbatim}
\begin{alltt}
\hlstd{geneUniverse} \hlkwb{<-} \hlstd{topTab}\hlopt{$}\hlstd{EntrezsA}
\hlkwd{length}\hlstd{(geneUniverse)}
\end{alltt}
\begin{verbatim}
## [1] 6218
\end{verbatim}
\begin{alltt}
\hlstd{writeGeneLists}\hlkwb{<-} \hlnum{FALSE}
\hlkwa{if}\hlstd{(writeGeneLists)\{}
  \hlkwd{write.csv}\hlstd{(geneListUp,} \hlkwc{file}\hlstd{=}\hlstr{"selectedAvsB.up.csv"}\hlstd{)}
  \hlkwd{write.csv}\hlstd{(geneListDown,} \hlkwc{file}\hlstd{=}\hlstr{"selectedAvsB.down.csv"}\hlstd{)}
  \hlkwd{write.csv}\hlstd{(geneUniverse,} \hlkwc{file}\hlstd{=}\hlstr{"geneUniverse.csv"}\hlstd{)}
\hlstd{\}}
\end{alltt}
\end{kframe}
\end{knitrout}

\section{Pathway Analysis}

We start by removing NA's (if any) and ensuring that we have unique identifiers.

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{# Remove potential NA's values}
\hlstd{geneEntrezsUp} \hlkwb{<-} \hlkwd{unique}\hlstd{(geneListUp[}\hlopt{!}\hlkwd{is.na}\hlstd{(geneListUp)])}
\hlstd{geneEntrezsDown} \hlkwb{<-} \hlkwd{unique}\hlstd{(geneListDown[}\hlopt{!}\hlkwd{is.na}\hlstd{(geneListDown)])}
\hlstd{geneEntrezsUniverse} \hlkwb{<-} \hlkwd{unique}\hlstd{(geneUniverse[}\hlopt{!}\hlkwd{is.na}\hlstd{(geneUniverse)])}
\end{alltt}
\end{kframe}
\end{knitrout}

We will use the \texttt{GOstats package} which proceeds in two steps:
\begin{enumerate}
  \item First we create the appropriate objects
  \item Next we use them to do the enrichment analysis
  \item In a final step we generate an html report with the test results
\end{enumerate}

First we create the appropriate objects
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{require}\hlstd{(GOstats)}
\hlcom{## Creamos los "hiperparametros" en que se basa el analisis}
\hlstd{GOparams} \hlkwb{=} \hlkwd{new}\hlstd{(}\hlstr{"GOHyperGParams"}\hlstd{,}
               \hlkwc{geneIds}\hlstd{=geneEntrezsUp,} \hlkwc{universeGeneIds}\hlstd{=geneEntrezsUniverse,}
               \hlkwc{annotation}\hlstd{=}\hlstr{"org.Hs.eg.db"}\hlstd{,} \hlcom{# might have use hgu133a.db instead}
               \hlkwc{ontology}\hlstd{=}\hlstr{"BP"}\hlstd{,}
               \hlkwc{pvalueCutoff}\hlstd{=}\hlnum{0.001}\hlstd{,} \hlkwc{conditional}\hlstd{=}\hlnum{FALSE}\hlstd{,}
               \hlkwc{testDirection}\hlstd{=}\hlstr{"over"}\hlstd{)}
\hlstd{KEGGparams} \hlkwb{=} \hlkwd{new}\hlstd{(}\hlstr{"KEGGHyperGParams"}\hlstd{,}
                 \hlkwc{geneIds}\hlstd{=geneEntrezsUp,} \hlkwc{universeGeneIds}\hlstd{=geneEntrezsUniverse,}
                 \hlkwc{annotation}\hlstd{=}\hlstr{"org.Hs.eg.db"}\hlstd{,} \hlcom{# might have use hgu133a.db instead}
                 \hlkwc{pvalueCutoff}\hlstd{=}\hlnum{0.01}\hlstd{,} \hlkwc{testDirection}\hlstd{=}\hlstr{"over"}\hlstd{)}
\end{alltt}
\end{kframe}
\end{knitrout}

Next we use them to do the enrichment analysis
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{GOhyper} \hlkwb{=} \hlkwd{hyperGTest}\hlstd{(GOparams)}
\hlstd{KEGGhyper} \hlkwb{=} \hlkwd{hyperGTest}\hlstd{(KEGGparams)}
\hlkwd{cat}\hlstd{(}\hlstr{"GO\textbackslash{}n"}\hlstd{)}
\end{alltt}
\begin{verbatim}
## GO
\end{verbatim}
\begin{alltt}
\hlkwd{print}\hlstd{(}\hlkwd{head}\hlstd{(}\hlkwd{summary}\hlstd{(GOhyper)))}
\end{alltt}
\begin{verbatim}
##       GOBPID         Pvalue OddsRatio ExpCount Count Size
## 1 GO:0000278 0.000000005457     2.585   27.989    60  479
## 2 GO:1903047 0.000000463897     2.406   24.308    50  416
## 3 GO:0007049 0.000000491386     2.006   47.856    81  819
## 4 GO:0048584 0.000002983857     1.842   58.257    91  997
## 5 GO:1905819 0.000013104020     8.247    1.578     9   27
## 6 GO:2000816 0.000013104020     8.247    1.578     9   27
##                                                         Term
## 1                                         mitotic cell cycle
## 2                                 mitotic cell cycle process
## 3                                                 cell cycle
## 4                positive regulation of response to stimulus
## 5               negative regulation of chromosome separation
## 6 negative regulation of mitotic sister chromatid separation
\end{verbatim}
\begin{alltt}
\hlkwd{cat}\hlstd{(}\hlstr{"KEGG\textbackslash{}n"}\hlstd{)}
\end{alltt}
\begin{verbatim}
## KEGG
\end{verbatim}
\begin{alltt}
\hlkwd{print}\hlstd{(}\hlkwd{head}\hlstd{(}\hlkwd{summary}\hlstd{(KEGGhyper)))}
\end{alltt}


{\ttfamily\noindent\itshape\color{messagecolor}{\#\# \\\#\# KEGG.db contains mappings based on older data because the original\\\#\#\ \  resource was removed from the the public domain before the most\\\#\#\ \  recent update was produced. This package should now be considered\\\#\#\ \  deprecated and future versions of Bioconductor may not have it\\\#\#\ \  available.\ \ Users who want more current data are encouraged to look\\\#\#\ \  at the KEGGREST or reactome.db packages}}\begin{verbatim}
##   KEGGID   Pvalue OddsRatio ExpCount Count Size
## 1  04110 0.001294     2.878    5.724    14   89
## 2  04114 0.002082     3.167    4.116    11   64
## 3  04914 0.002461     3.590    3.023     9   47
## 4  04010 0.004909     2.352    7.267    15  113
## 5  04062 0.006140     2.452    6.045    13   94
## 6  04971 0.007421     4.082    1.801     6   28
##                                      Term
## 1                              Cell cycle
## 2                          Oocyte meiosis
## 3 Progesterone-mediated oocyte maturation
## 4                  MAPK signaling pathway
## 5             Chemokine signaling pathway
## 6                  Gastric acid secretion
\end{verbatim}
\end{kframe}
\end{knitrout}

In a final step we generate an html report with the test results

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{# Creamos un informe html con los resultados}
\hlstd{GOfilename} \hlkwb{=}\hlkwd{file.path}\hlstd{(}\hlkwd{paste}\hlstd{(}\hlstr{"GOResults.AvsB.up"}\hlstd{,}\hlstr{".html"}\hlstd{,} \hlkwc{sep}\hlstd{=}\hlstr{""}\hlstd{))}
\hlstd{KEGGfilename} \hlkwb{=}\hlkwd{file.path}\hlstd{(}\hlkwd{paste}\hlstd{(}\hlstr{"KEGGResults.AvsB.up"}\hlstd{,}\hlstr{".html"}\hlstd{,} \hlkwc{sep}\hlstd{=}\hlstr{""}\hlstd{))}
\hlkwd{htmlReport}\hlstd{(GOhyper,} \hlkwc{file} \hlstd{= GOfilename,} \hlkwc{summary.args}\hlstd{=}\hlkwd{list}\hlstd{(}\hlstr{"htmlLinks"}\hlstd{=}\hlnum{TRUE}\hlstd{))}
\hlkwd{htmlReport}\hlstd{(KEGGhyper,} \hlkwc{file} \hlstd{= KEGGfilename,} \hlkwc{summary.args}\hlstd{=}\hlkwd{list}\hlstd{(}\hlstr{"htmlLinks"}\hlstd{=}\hlnum{TRUE}\hlstd{))}
\end{alltt}
\end{kframe}
\end{knitrout}

\section{Analysis of Functional Profiles}

The \texttt{goProfiles} package provides a different approach to Pathway Analysis. Its most distinctive characteristic is the possibility of projecting gene lists on ``levels" of the Gene Ontology and compare these projections between lists (compare lists based on their projections).

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{require}\hlstd{(goProfiles)}
\end{alltt}


{\ttfamily\noindent\itshape\color{messagecolor}{\#\# Loading required package: goProfiles}}

{\ttfamily\noindent\itshape\color{messagecolor}{\#\# Loading required package: CompQuadForm}}

{\ttfamily\noindent\itshape\color{messagecolor}{\#\# Loading required package: stringr}}

{\ttfamily\noindent\itshape\color{messagecolor}{\#\# \\\#\# Attaching package: 'stringr'}}

{\ttfamily\noindent\itshape\color{messagecolor}{\#\# The following object is masked from 'package:graph':\\\#\# \\\#\#\ \ \ \  boundary}}\begin{alltt}
\hlstd{BPprofile1}\hlkwb{<-} \hlkwd{basicProfile}\hlstd{(}\hlkwc{genelist}\hlstd{=geneListUp,} \hlkwc{onto}\hlstd{=}\hlstr{"BP"}\hlstd{,} \hlkwc{orgPackage}\hlstd{=}\hlstr{"org.Hs.eg.db"}\hlstd{,} \hlkwc{empty.cats}\hlstd{=}\hlnum{FALSE}\hlstd{,} \hlkwc{level}\hlstd{=}\hlnum{2}\hlstd{)[[}\hlnum{1}\hlstd{]]}
\hlkwd{head}\hlstd{(BPprofile1)}
\end{alltt}
\begin{verbatim}
##              Description       GOID Frequency
## 8               behavior GO:0007610        10
## 15   biological adhesion GO:0022610        36
## 21      biological phase GO:0044848         1
## 29 biological regulation GO:0065007       271
## 3           cell killing GO:0001906         5
## 9     cell proliferation GO:0008283        82
\end{verbatim}
\end{kframe}
\end{knitrout}

Now we might want to annotate the GO categories with their genes.
First we look the reverse, which GO terms are associated with each gene in the list

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{require}\hlstd{(org.Hs.eg.db)}
\hlkwd{keytypes}\hlstd{(org.Hs.eg.db)}
\end{alltt}
\begin{verbatim}
##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GO"           "GOALL"        "IPI"          "MAP"          "OMIM"        
## [16] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"       "UNIGENE"     
## [26] "UNIPROT"
\end{verbatim}
\begin{alltt}
\hlstd{entrezsUp2GO} \hlkwb{<-} \hlkwd{select}\hlstd{(org.Hs.eg.db,} \hlkwc{keys} \hlstd{=} \hlkwd{as.character}\hlstd{(geneListUp),} \hlkwc{columns}\hlstd{=}\hlkwd{c}\hlstd{(}\hlstr{"SYMBOL"}\hlstd{,} \hlstr{"GOALL"}\hlstd{))}
\end{alltt}


{\ttfamily\noindent\itshape\color{messagecolor}{\#\# 'select()' returned 1:many mapping between keys and columns}}\begin{alltt}
\hlkwd{head}\hlstd{(entrezsUp2GO)}
\end{alltt}
\begin{verbatim}
##   ENTREZID SYMBOL      GOALL EVIDENCEALL ONTOLOGYALL
## 1    51442  VGLL1 GO:0000981         IBA          MF
## 2    51442  VGLL1 GO:0003674         IBA          MF
## 3    51442  VGLL1 GO:0003674         TAS          MF
## 4    51442  VGLL1 GO:0003700         IBA          MF
## 5    51442  VGLL1 GO:0003705         IBA          MF
## 6    51442  VGLL1 GO:0003712         TAS          MF
\end{verbatim}
\begin{alltt}
\hlstd{entrezsUp2GOBP}\hlkwb{<-} \hlstd{entrezsUp2GO[entrezsUp2GO}\hlopt{$}\hlstd{ONTOLOGY}\hlopt{==}\hlstr{"BP"}\hlstd{,]}
\hlstd{BPprofileWithGenes}\hlkwb{<-} \hlkwd{cbind}\hlstd{(BPprofile1,} \hlkwc{genes}\hlstd{=}\hlkwd{rep}\hlstd{(}\hlstr{""}\hlstd{,} \hlkwd{nrow}\hlstd{(BPprofile1)))}
\hlstd{BPprofileWithGenes}\hlopt{$}\hlstd{genes}\hlkwb{<-} \hlkwd{as.character}\hlstd{(BPprofileWithGenes}\hlopt{$}\hlstd{genes)}
\hlkwa{for} \hlstd{(i} \hlkwa{in} \hlnum{1}\hlopt{:}\hlkwd{nrow}\hlstd{(BPprofile1))\{}
  \hlstd{GOIDi}\hlkwb{<-} \hlstd{BPprofile1[i,}\hlstr{"GOID"}\hlstd{]}
  \hlstd{genesi} \hlkwb{<-}\hlkwd{unique}\hlstd{(entrezsUp2GOBP[entrezsUp2GOBP}\hlopt{$}\hlstd{GO}\hlopt{==}\hlstd{GOIDi,}\hlstr{"ENTREZID"}\hlstd{])}
  \hlstd{genesi} \hlkwb{<-} \hlkwd{paste}\hlstd{(genesi[}\hlopt{!}\hlkwd{is.na}\hlstd{(genesi)],} \hlkwc{collapse} \hlstd{=} \hlstr{" "}\hlstd{)}
  \hlstd{BPprofileWithGenes[i,}\hlstr{"genes"}\hlstd{]}\hlkwb{=}\hlstd{genesi}
\hlstd{\}}
\hlkwd{head}\hlstd{(BPprofileWithGenes)}
\end{alltt}
\begin{verbatim}
##              Description       GOID Frequency
## 8               behavior GO:0007610        10
## 15   biological adhesion GO:0022610        36
## 21      biological phase GO:0044848         1
## 29 biological regulation GO:0065007       271
## 3           cell killing GO:0001906         5
## 9     cell proliferation GO:0008283        82
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             genes
## 8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        3251 2744 6566 6722 2019 51083 3815 5743 89797 2534 5270
## 15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       5613 1029 2824 9655 6347 4851 6722 23189 57162 5317 4155 3815 10892 11151 10125 5411 10855 5771 2213 4478 6777 5054 1326 2534 1230 4690 8874 4067 5578 119 3569 5062 5270 965 9473 79679
## 21                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          51083
## 29 51442 2568 5613 1029 3251 10644 2171 2744 81611 467 23708 6566 2824 113 51444 7447 4281 3843 7298 4609 53335 9126 30968 4603 4893 2037 9655 663 9459 23228 9493 6347 28959 2773 5359 5268 4200 9833 55795 7153 4851 7027 858 3832 6001 51765 6663 146909 11326 10099 8428 6722 55872 10000 23643 10950 80306 10591 8842 1116 65109 9444 23189 9181 5321 10477 2019 699 10449 57162 231 9232 10149 10659 51083 929 10133 10479 56603 5317 54908 7272 2296 114885 2263 4155 6275 23432 2305 6782 4277 10492 9994 10112 2332 5366 55013 5213 4751 9446 597 5066 55704 6624 7342 10635 26278 6197 10915 8660 3815 6194 5520 55165 55144 64425 10892 55120 8881 29899 8190 5806 11151 10125 84079 26577 6355 79675 8685 8776 86 7913 8898 22873 9839 5202 57180 1871 6907 3709 10855 29969 10670 51389 26098 403 5771 27042 10289 5743 2213 9925 8936 10308 8877 6374 5585 3868 5134 4478 1808 64783 1514 4311 64699 123 1063 7096 6711 55365 55055 5984 55816 89797 3455 10966 3759 6777 6877 5054 1326 2001 834 1381 65110 10602 51053 5996 10484 7159 51320 5610 5292 23390 9467 2534 9910 4089 701 259266 4332 1230 2212 4690 8874 2210 2869 573 54962 6742 9656 10542 54861 79001 4067 54210 2146 5578 10054 440 366 119 6201 9933 10346 9133 1756 8477 3925 29919 3569 4436 6398 85453 6840 23683 10209 56647 5062 56652 3837 23076 9702 6277 5270 2619 5650 2769 965 56950 9473 79679 996 10403 29978 8795 23347 991 3298 56983 5329 55012 9045 1535 1054 374354 10404 54780 2919 23516 7775
## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            3251 4277 11151 10125 2213 6777 1054
## 9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            5613 1029 3251 467 4609 4893 23228 6347 2773 5268 9833 4851 7027 858 6663 6722 10000 10950 10591 9181 5321 699 57162 231 51083 7272 2296 2263 2305 55704 6624 9837 2012 8660 3815 6194 10892 55120 8881 8190 11151 6355 9839 1871 56992 10855 5771 5743 2213 8877 6374 5585 3868 5134 4478 1063 6777 2001 5610 5292 2534 4089 701 259266 4332 4690 2869 54962 4067 2146 5578 9133 3569 85453 6277 5270 56950 79679 996 991 1535 2919
\end{verbatim}
\end{kframe}
\end{knitrout}




\end{document}
